src/Angor/Shared/Services/RelayService.cs

```diff
@@ -89,6 +89,58 @@ public void RequestProjectCreateEventsByPubKey(Action<NostrEvent> onResponseActi
                 Kinds = [NostrKind.ApplicationSpecificData, NostrKind.Metadata],
             }));
         }
+        
+        // add the same as function ^^ as async
+        public async Task RequestProjectCreateEventsByPubKeyAsync(
+            string[] nPubs,
+            Action<NostrEvent> onResponseAction,
+            Action? onEoseAction)
+        {
+            var subscriptionKey = Guid.NewGuid().ToString().Replace("-", "");
+
+            var nostrClient = _communicationFactory.GetOrCreateClient(_networkService);
+
+            if (!_subscriptionsHandling.RelaySubscriptionAdded(subscriptionKey))
+            {
+                var subscription = nostrClient.Streams.EventStream
+                    .Where(_ => _.Subscription == subscriptionKey)
+                    .Where(_ => _.Event is not null)
+                    .Select(_ => _.Event)
+                    .Subscribe(onResponseAction!);
+
+                _subscriptionsHandling.TryAddRelaySubscription(subscriptionKey, subscription);
+            }
+
+            if (onEoseAction != null)
+            {
+                _subscriptionsHandling.TryAddEoseAction(subscriptionKey, onEoseAction);
+            }
+
+            var tcs = new TaskCompletionSource();
+
+            // Handle end-of-stream asynchronously
+            if (onEoseAction == null)
+            {
+                _subscriptionsHandling.TryAddEoseAction(subscriptionKey, () => tcs.SetResult());
+            }
+            else
+            {
+                _subscriptionsHandling.TryAddEoseAction(subscriptionKey, () =>
+                {
+                    onEoseAction();
+                    tcs.SetResult();
+                });
+            }
+
+            nostrClient.Send(new NostrRequest(subscriptionKey, new NostrFilter
+            {
+                Authors = nPubs,
+                Kinds = new[] { NostrKind.ApplicationSpecificData, NostrKind.Metadata },
+            }));
+
+            await tcs.Task; // Wait for the end of stream
```

Comment:
Why are we awaiting the tcs and not just allowing the callback thread to perfome the action we need?

Reply:
I did it because of this 
https://learn.microsoft.com/en-us/dotnet/standard/asynchronous-programming-patterns/implementing-the-task-based-asynchronous-pattern

why shouldnt we need `TaskCompletionSource ` to wait for end of steam?

either way copilot removed it 😉